Bring gdb stub up-to-date for user programs

sanderssj opened the pull request:
This change updates the in-kernel gdb stub to work again with multi-threaded
user programs running on multiple cpus. It breaks out the architecture
specific parts to allow for future ARM debugging support. This adds support
for thread commands, status, and id. The gdb exception handler was updated
to pass through certain exceptions like page faults, and the handler is
now installed for all frames of all threads on the user process.

Breakpoint handling has been updated allowing hardware breakpoints
to be installed on all cpus, and using the correct code of hardware
breakpoints. This allows the use of software breakpoints which are much
less limiting for line-by-line debugging. Software breaks are not
implemented directly, but deferred to gdb so it will manually insert
trap instructions. Text pages are force updated with write permission
to allow traps to be inserted, although it's not a long term solution.

A new queue is added to the cpuinfo struct to allow thunks to be run
for that specific cpu. This is necessary for hardware breakpoints to
be modified on every cpu. This should also be temporary until thunks
can be scheduled per cpu.

There is further work to be done on the stub so this should be
considered a work-in-progress update with expected limits and gaps
in stub functionality.

wjhun reviewed (approved):
I made some minor comments, but otherwise everything looks good.

wjhun commented:
There is a pre-existing issue in gdbtcp.c that is worth mentioning now. A gdb exception could lead to a call to putpacket, which would in turn apply the output_handler (currently only gdb_send). This would call tcp_write at exception level, which is not safe. Perhaps it could instead enqueue the output and schedule a service routine, if not already scheduled, to the runqueue. Even a spinlock protecting g->output could suffice as an output queue, I think.

sanderssj commented on src/runtime/text.h:
```diff
@@ -169,6 +169,21 @@ static inline boolean parse_int(buffer b, u32 base, u64 *result)
+{
+  int sign = 1;
+
+  if (*(u8 *)buffer_ref(b, 0) == '-') {
```
Good point, added in a check!

sanderssj commented on src/unix/exec.c:
```diff
@@ -142,8 +142,8 @@ void start_process(thread t, void *start)
-        rputs("TODO: in-kernel gdb needs revisiting\n");
-//        init_tcp_gdb(heap_general(get_kernel_heaps()), t->p, 9090);
+        rputs("NOTE: in-kernel gdb is a work in progress\n");
+        init_tcp_gdb(heap_general(get_kernel_heaps()), t->p, 9090);
```
Good catch, fixed.

sanderssj commented on src/gdb/gdbstub.c:
```diff
@@ -177,129 +282,192 @@ static boolean handle_request(gdb g, buffer b, buffer output)
+                    if (breakpoint_insert(g->h, addr, 0, 8, closure(g->h, send_ok, g)))
+                        return false;
                     else
                         bprintf(output, "E08");
```
Added

sanderssj commented on src/gdb/gdbstub.c:
```diff
@@ -77,55 +145,74 @@ static struct handler query_handler[] = {
-    // a little more symbolic here please
-    if (stepping) {
-        thread_frame(g->t)[FRAME_FLAGS] |= TRAP_FLAG;
+    if (t == INVALID_ADDRESS && thread_from_tid(g->p, g->ctid) != INVALID_ADDRESS) {
```
The intent here was muddy, so I've refactored it to be a little more obvious and eliminated the extra thread_from_tid.

sanderssj commented:
> There is a pre-existing issue in gdbtcp.c that is worth mentioning now. A gdb exception could lead to a call to putpacket, which would in turn apply the output_handler (currently only gdb_send). This would call tcp_write at exception level, which is not safe. Perhaps it could instead enqueue the output and schedule a service routine, if not already scheduled, to the runqueue. Even a spinlock protecting g->output could suffice as an output queue, I think.

Good catch, I have added a way to defer the response from the exception handler. I ended up protecting g->send_buffer which is what gets copied in the network code and is a little simpler. I updated the handler to use a stack buffer for
the response as it has a small fixed maximum size so there's no concurrency issue between it and the normal command response.
